modules/core/src/main/java/org/apache/ignite/internal/processors/cache/ExchangeContext.java

```diff
@@ -100,6 +106,13 @@ public boolean fetchAffinityOnJoin() {
         return fetchAffOnJoin;
     }
 
+    /**
+     * @return {@code True} if it's safe to perform PME-free switch.
+     */
+    public boolean exchangeFreeSwitch() {
```

Comment:
Should we consider this PME non-blocking cache operations? If so, we need to fix the metric (see `GridDhtPartitionsExchangeFuture#currentPMEDuration`).

Reply:
Since exchange-free switch blocks upcoming operations during the tx recovery, seems, it's correct to consider it's blocking during this period.
In other words exchange-free does not means non-blocking, but period is short
Blocking period can be shortened in future by recovery procedure refactoring.
Also we, theoretically, can block only affected nodes (we have some cornercases here, so it should be additional fix)